truenas_installer/dialog.py

```diff
@@ -0,0 +1,136 @@
+import os
+
+import textwrap
+
+import asyncio
+import subprocess
+import tempfile
+
+__all__ = ["dialog", "dialog_checklist", "dialog_menu", "dialog_msgbox", "dialog_yesno"]
+
+
+async def dialog(args, check=False):
+    args = ["dialog"] + args
+
+    process = await asyncio.create_subprocess_exec(*args, stderr=subprocess.PIPE)
+    _, stderr = await process.communicate()
+
+    stderr = stderr.decode("utf-8", "ignore")
+
+    if check:
+        if process.returncode != 0:
+            raise subprocess.CalledProcessError(process.returncode, args, stderr=stderr)
+
+    return subprocess.CompletedProcess(args, process.returncode, stderr=stderr)
+
+
+async def dialog_checklist(title, text, items):
+    result = await dialog(
+        [
+            "--clear",
+            "--title", title,
+            "--checklist", text, "20", "60", "0"
+        ] +
+        sum(
+            [
+                [k, v, "off"]
+                for k, v in items.items()
+            ],
+            [],
+        )
+    )
+
+    if result.returncode == 0:
+        return result.stderr.split()
+    else:
+        return None
+
+
+async def dialog_menu(title, items):
+    result = await dialog(
+        [
+            "--clear",
+            "--title", title,
+            "--menu", "", "12", "73", "6"
+        ] +
+        sum(
+            [
+                [str(i + 1), title]
+                for i, title in enumerate(items.keys())
+            ],
+            [],
+        )
+    )
+
+    if result.returncode == 0:
+        return await list(items.values())[int(result.stderr) - 1]()
+    else:
+        return None
+
+
+async def dialog_msgbox(title, text):
+    await dialog([
+        "--clear",
+        "--title", title,
+        "--msgbox", text,
+        str(4 + len(text.rstrip().splitlines())), "60",
+    ])
+
+
+async def dialog_password(title):
+    with tempfile.NamedTemporaryFile("w") as dialogrc:
```

Comment:
I'm noticing a trend where all of these are coroutines but we're doing a ton of blocking I/O. I understand this is an installer and it's synchronous by design but it doesn't sit right with me to break the asyncio best practices. Seems sloppy.

The asyncio base library has a `to_thread` coroutine that makes it pretty easy and clean to call blocking I/O.